### ISSUE TITLE
New connections from Nginx log localhost as IP address

### ISSUE THREAD

**Issue by collinhundley** [IBM-Swift/Kitura#876]
When running Kitura behind Nginx (using `proxy_pass`), all new connections are logged like this:

```sh
Accepted connection from: 127.0.0.1:39462 
```

Logging localhost for the remote IP isn't very useful. Is there a way we can get the original IP address when connections come from a rProxy?  I don't know if the same thing happens with Apache.

**Comment by na-gupta** [comment]
You can add the real remote address as extra headers using ngnix embedded variables in your nginx config. For example: 
```
proxy_set_header   X-Real-IP   $remote_addr;
proxy_set_header   X-Forwarded-For   $proxy_add_x_forwarded_for;
```

Then you can get the values in Kitura using `request.headers["X-Forwarded-For"]`

**Comment by collinhundley** [comment]
@na-gupta that works great for getting it through the header, but it still doesn't solve the logger problem. It looks like the address gets logged when an instance of `Socket` is first created, so I'm not sure the headers would be available at that point.

**Comment by na-gupta** [comment]
Yes, the application would need to log it using a custom router handler.

**Comment by collinhundley** [comment]
That's understandable, but in my opinion it's an issue that the logs are littered with thousands (or potentially millions) of lines of useless information. You would expect a "new connection" log to contain the real IP address of the client, but that's not the case for the 95% of people who run their servers behind a reverse proxy.

**Comment by na-gupta** [comment]
True, but I don't think Kitura code should use custom headers for this purpose. You could always set the logging level to info or warning so these verbose level client connection messages don't clutter the logs. Do you have an alternate suggestion for improvement that would make this more generic and useful?

**Comment by collinhundley** [comment]
Well `X-Forwarded-For` is a standard header, I wouldn't really call it "custom." However, I'm not sure I see a solution at the moment. Logging new connections at the socket level could be useful for low-level debugging, but causes clutter for some applications. Delaying the log until after the headers are parsed is one option, but could potentially cause confusion if an error occurred before that happened.

**Comment by na-gupta** [comment]
We are thinking about changing the existing log line at the socket layer to **debug** (lower priority than verbose) and adding a new **verbose** log line after we parse the headers which checks the new **Forwarded:** or older **X-Forwarded-For:** and **X-Forwarded-Proto:** headers and logs it accordingly. Perhaps something like:

When no forwarded headers present:
`http(s)/fastcgi request from: <socket.host>`

When new **Forwarded** header is present, use that in preference to **X-Forwarded-For**
`http(s)/fastcgi request: Forwarded: "<header_value>"`
which should translate to:
`http(s)/fastcgi request: Forwarded: "for=<client_ip>; proto=http(s); by=<proxy_ip>"`

Otherwise, when **X-Forwarded-For** header is present:
`http/fastcgi request: X-Forwarded-For: <header_value> by <socket.host>`
which should translate to:
`http/fastcgi request: X-Forwarded-For: <client_ip> by <proxy_ip>`

@shmuelk @collinhundley Any thoughts?

**Comment by collinhundley** [comment]
@na-gupta that seems like a good solution to me. I'm assuming there's a place where we have access to the headers and the original socket info at the same time?

**Comment by na-gupta** [comment]
@collinhundley yes, I think there is. Will put in this change in the next day or two.

**Comment by na-gupta** [comment]
@collinhundley This is now available in Kitura-1.3.0
